### PR TITLE
Scatter batching rule (closes #350)

### LINKED ISSUE #350: lax.scatter batching (vmap) rule
Follow-up from #246.

### PR DESCRIPTION
**Pull request by mattjj**
This PR includes commits from #351, but those will get squashed away when we merge #351.

The main use case is vmap(grad(indexing_fun)).

### CODE REVIEW

**Comment by mattjj** [comment]
@hawkinsp the unimplemented case with the TODO(mattjj,phawkins) on it can't be generated by vmap+grad on lax.gather, but only by calling lax.scatter_add directly (and vmapping over it).

File: jax/lax.py [hunk 1]
```diff
@@ -2233,12 +2232,63 @@ def _scatter_transpose_rule(t, operand, scatter_indices, updates,
...
+  del update_jaxpr, update_consts, updates_shape  # Unused.
+
+  size = next(x.shape[ax] for x, ax in zip(batched_args, batch_dims)
+              if ax is not None)
+  operand = batching.bdim_at_front(operand, operand_bdim, broadcast_size=size,
+                                   force_broadcast=True)
+  operand_bdim = 0
+
+  if scatter_indices_bdim is not None and updates_bdim is None:
+    raise NotImplementedError  # TODO(mattjj,phawkins)
```
Comment (hawkinsp) [inline jax/lax.py]: Maybe do this too, given you already did the harder case?

File: jax/lax.py [hunk 2]
```diff
@@ -2233,12 +2232,63 @@ def _scatter_transpose_rule(t, operand, scatter_indices, updates,
...
 
+def _scatter_batching_rule(batched_args, batch_dims, update_jaxpr,
+                           update_consts, dimension_numbers, updates_shape):
+  operand, scatter_indices, updates = batched_args
+  operand_bdim, scatter_indices_bdim, updates_bdim = batch_dims
+  del update_jaxpr, update_consts, updates_shape  # Unused.
+
+  size = next(x.shape[ax] for x, ax in zip(batched_args, batch_dims)
+              if ax is not None)
+  operand = batching.bdim_at_front(operand, operand_bdim, broadcast_size=size,
```
Comment (hawkinsp) [inline jax/lax.py]: I don't follow this logic. Add a comment?

Comment (mattjj) [reply jax/lax.py]: Done!

File: jax/lax.py (hunk 1 above)
Comment (mattjj) [reply jax/lax.py]: Nah, I want to move on to other things and it's not clear that we care about this other case (which can't be generated by vmap+grad on gather, but only by vmapping over lax.scatter_add directly).

**Pull request merged**